### PR TITLE
Hipdnn integration and batchnorm

### PR DESCRIPTION
**Pull request by dnikolaev-amd**
## Motivation

Use HipDNN in Pytorch

## Technical Details

Requirements:
- ROCm 7.12 or higher

Main features:
- Integrate HipDNN with Pytorch if available (cmake, includes, libs, etc.)
- Use HipDNN for batchnorm if available

What was done:
- Cmake definition `USE_HIPDNN` if HipDNN is available
- Runtime hook `CUDAHooks::compiledWithHipDNN()`  
- Add Pytorch `HipdnnModule` to handle related settings in Pytorch
- Add `torch.backends.hipdnn.is_available()` flag to check if HipDNN currently available in python
- Add `torch.backends.hipdnn.flags(enable=None)` to globalContext() (currently OFF by default until it stabilizes, MIOpen backend will be used)
- `TEST_HIPDNN` boolean value to enable/disable related python UTs
- Add `BatchNormBackend::Hipdnn` to the batchnorm backends enum and new test to validate backend selection `test_nn.py::TestNN::test_batchnorm_hipdnn_backend_selection`
- Add new tests `test_nn.py::TestNN::test_batchnorm_hipdnn_2D_*` and `3D` if  HipDNN available:
> ```
> test_batchnorm_hipdnn_2D_inference_NCHW_vs_cpu_float32
> test_batchnorm_hipdnn_2D_inference_NCHW_vs_cpu_mixed_bfloat16
> test_batchnorm_hipdnn_2D_inference_NCHW_vs_cpu_mixed_float16
> test_batchnorm_hipdnn_2D_inference_NCHW_vs_native_float32
> test_batchnorm_hipdnn_2D_inference_NCHW_vs_native_mixed_bfloat16
> test_batchnorm_hipdnn_2D_inference_NCHW_vs_native_mixed_float16
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_NCHW_float32
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_NCHW_mixed_bfloat16
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_NCHW_mixed_float16
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_cpu_float32
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_cpu_mixed_bfloat16
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_cpu_mixed_float16
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_native_float32
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_native_mixed_bfloat16
> test_batchnorm_hipdnn_2D_inference_NHWC_vs_native_mixed_float16
> test_batchnorm_hipdnn_2D_train_NCHW_vs_cpu_float32
> test_batchnorm_hipdnn_2D_train_NCHW_vs_cpu_mixed_bfloat16
> test_batchnorm_hipdnn_2D_train_NCHW_vs_cpu_mixed_float16
> test_batchnorm_hipdnn_2D_train_NCHW_vs_native_float32
> test_batchnorm_hipdnn_2D_train_NCHW_vs_native_mixed_bfloat16
> test_batchnorm_hipdnn_2D_train_NCHW_vs_native_mixed_float16
> test_batchnorm_hipdnn_2D_train_NHWC_vs_NCHW_float32
> test_batchnorm_hipdnn_2D_train_NHWC_vs_NCHW_mixed_bfloat16
> test_batchnorm_hipdnn_2D_train_NHWC_vs_NCHW_mixed_float16
> test_batchnorm_hipdnn_2D_train_NHWC_vs_cpu_float32
> test_batchnorm_hipdnn_2D_train_NHWC_vs_cpu_mixed_bfloat16
> test_batchnorm_hipdnn_2D_train_NHWC_vs_cpu_mixed_float16
> test_batchnorm_hipdnn_2D_train_NHWC_vs_native_float32
> test_batchnorm_hipdnn_2D_train_NHWC_vs_native_mixed_bfloat16
> test_batchnorm_hipdnn_2D_train_NHWC_vs_native_mixed_float16
> test_batchnorm_hipdnn_3D_inference_NCHW_vs_cpu_float32
> test_batchnorm_hipdnn_3D_inference_NCHW_vs_cpu_mixed_bfloat16
> test_batchnorm_hipdnn_3D_inference_NCHW_vs_cpu_mixed_float16
> test_batchnorm_hipdnn_3D_inference_NCHW_vs_native_float32
> test_batchnorm_hipdnn_3D_inference_NCHW_vs_native_mixed_bfloat16
> test_batchnorm_hipdnn_3D_inference_NCHW_vs_native_mixed_float16
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_NCHW_float32
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_NCHW_mixed_bfloat16
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_NCHW_mixed_float16
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_cpu_float32
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_cpu_mixed_bfloat16
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_cpu_mixed_float16
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_native_float32
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_native_mixed_bfloat16
> test_batchnorm_hipdnn_3D_inference_NHWC_vs_native_mixed_float16
> test_batchnorm_hipdnn_3D_train_NCHW_vs_cpu_float32
> test_batchnorm_hipdnn_3D_train_NCHW_vs_cpu_mixed_bfloat16
> test_batchnorm_hipdnn_3D_train_NCHW_vs_cpu_mixed_float16
> test_batchnorm_hipdnn_3D_train_NCHW_vs_native_float32
> test_batchnorm_hipdnn_3D_train_NCHW_vs_native_mixed_bfloat16
> test_batchnorm_hipdnn_3D_train_NCHW_vs_native_mixed_float16
> test_batchnorm_hipdnn_3D_train_NHWC_vs_NCHW_float32
> test_batchnorm_hipdnn_3D_train_NHWC_vs_NCHW_mixed_bfloat16
> test_batchnorm_hipdnn_3D_train_NHWC_vs_NCHW_mixed_float16
> test_batchnorm_hipdnn_3D_train_NHWC_vs_cpu_float32
> test_batchnorm_hipdnn_3D_train_NHWC_vs_cpu_mixed_bfloat16
> test_batchnorm_hipdnn_3D_train_NHWC_vs_cpu_mixed_float16
> test_batchnorm_hipdnn_3D_train_NHWC_vs_native_float32
> test_batchnorm_hipdnn_3D_train_NHWC_vs_native_mixed_bfloat16
> test_batchnorm_hipdnn_3D_train_NHWC_vs_native_mixed_float16
>```

## Test Plan

`HIPDNN_LOG_LEVEL=info` environment variable can be used to enable HipDNN console logging to make sure it is used

```bash
PYTORCH_TEST_WITH_ROCM=1 python test/test_nn.py -v TestNN.test_batchnorm_hipdnn_backend_selection \
TestNN.test_batchnorm_hipdnn_2D_train_NCHW_vs_cpu_float32 \
TestNN.test_batchnorm_hipdnn_3D_inference_NHWC_vs_cpu_float32 \
TestNN.test_batchnorm_hipdnn_2D_train_NCHW_vs_cpu_mixed_float16 \
TestNN.test_batchnorm_hipdnn_3D_inference_NHWC_vs_cpu_mixed_bfloat16
```

Test Result:

```bash
test_batchnorm_hipdnn_backend_selection ... ok
test_batchnorm_hipdnn_2D_train_NCHW_vs_cpu_float32 ... ok
test_batchnorm_hipdnn_3D_train_NHWC_vs_cpu_float32 ... ok
test_batchnorm_hipdnn_2D_train_NCHW_vs_cpu_mixed_float16 ... ok
test_batchnorm_hipdnn_3D_train_NHWC_vs_cpu_mixed_bfloat16 ... ok
----------------------------------------------------------------------
Ran 5 tests in 0.259s
OK
```

### CODE REVIEW

**Comment by dnikolaev-amd** [comment]
@zjgarvey

**Review by zjgarvey — APPROVED** [review]
LGTM

**Review by zjgarvey — APPROVED** [review]
